File: leetcode/python/16.3SumClosest.py
import sys

class Test:

  @staticmethod
  def threeSumClosest(nums,target):
    if len(nums) == 0:
      return 0    
    nums.sort()
    # python 2 -> sys.maxint
    # python 3 > sys.maxsize
    minimum =  sys.maxsize
    closed = 0

    for i in range(len(nums)):
      if i == 0 or nums[i] != nums[i-1]:
        start = i+1
        end = len(nums) - 1
        while start < end :
          s = nums[i] + nums[start] + nums[end]
          if abs(s-target)<minimum:
            minimum = abs(s-target)
            closed = s
          if s < target:
            start+=1
          else:
            end-=1
    return closed
            
print(Test.threeSumClosest([-1, 0, 1, 2, -1, -4],8))

File: leetcode/js/16.3SumClosest.js
class Test{

    static threeSumClosed(nums,target){
      if(nums.length==0){
        return 0
      }
      nums.sort((a,b)=>a-b)
      let minimum = Number.MAX_VALUE
      // Number.MIN_SAFE_INTEGER
      var closed = 0 
      for(let i=0;i<nums.length;i++){
        if(i==0||nums[i]!=nums[i-1]){
          let start = i+1
          let end = nums.length - 1
          while(start<end){
            let s = nums[i] + nums[start] + nums[end]
            if(Math.abs(s-target)<minimum){
              minimum = Math.abs(s-target)
              closed =s 
            }
            if(s<target){start+=1}
            else{end-=1}
            }
          }
        }
          return closed
      }
    }
    
    
    console.log(Test.threeSumClosed([-1, 0, 1, 2, -1, -4],8))